lib/web_ui/lib/src/engine/semantics/heading.dart

```diff
@@ -3,35 +3,44 @@
 // found in the LICENSE file.
 
 import '../dom.dart';
+import 'label_and_value.dart';
 import 'semantics.dart';
 
 /// Renders semantics objects as headings with the corresponding
 /// level (h1 ... h6).
 class Heading extends PrimaryRoleManager {
   Heading(SemanticsObject semanticsObject)
       : super.blank(PrimaryRole.heading, semanticsObject) {
-        addHeadingRole();
-      }
-
-  @override
-  void update() {
-    super.update();
-
-    if (!semanticsObject.isHeadingLevelDirty) {
-      return;
-    }
-
-    addHeadingLevel(semanticsObject.headingLevel);
+    addFocusManagement();
+    addLiveRegion();
+    addRouteName();
+    addLabelAndValue(preferredRepresentation: LabelRepresentation.domText);
   }
 
   @override
-  bool focusAsRouteDefault() => focusable?.focusAsRouteDefault() ?? false;
-
-  void addHeadingRole() {
-    setAriaRole('heading');
-  }
+  DomElement createElement() => createDomElement('h${semanticsObject.headingLevel}');
```

Comment:
What happens if the `headingLevel` changes? Does the update automatically create a new element and call this method?

Reply:
Heading level should not change. It confuses screen readers. Web crawlers snapshot the heading at its default value too, so changing has no effect either. We could issue a warning, but I'd rather such warnings are issued by the framework, and the engine will just do a graceful degradation of functionality (in this case the heading level stays the same forever).